pkg/status/singletonstatus.go

```diff
@@ -0,0 +1,223 @@
+		for _, item := range bdg.Spec.Workload.NamespaceScope {
+			nsWObjRef := item.NamespaceScopeDownsyncObject
+			lister, found := c.listers.Get(schema.GroupVersionResource(nsWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", nsWObjRef.GroupVersionResource)
+			}
+			nsWObj, err := lister.ByNamespace(nsWObjRef.Namespace).Get(nsWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := nsWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: nsWObjRef.Group, Version: nsWObjRef.Version, Kind: nsWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   nsWObjRef.Resource,
+					ObjectName: cache.NewObjectName(nsWObjRef.Namespace, nsWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should have status synced", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
+					c.singletonState.set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object status syncing is not driven by Binding", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName, "binding", bdg.Name)
+					c.singletonState.addIfNotExist(wObjIdentifier)
+				}
+			}
+		}
+		for _, item := range bdg.Spec.Workload.ClusterScope {
+			cWObjRef := item.ClusterScopeDownsyncObject
+			lister, found := c.listers.Get(schema.GroupVersionResource(cWObjRef.GroupVersionResource))
+			if !found {
+				return fmt.Errorf("could not get lister for gvr: %s", cWObjRef.GroupVersionResource)
+			}
+			clusterWObj, err := lister.Get(cWObjRef.Name)
+			if err != nil {
+				return err
+			}
+			labels := clusterWObj.(metav1.Object).GetLabels()
+			if v, ok := labels[util.BindingPolicyLabelSingletonStatusKey]; ok {
+				wObjIdentifier := util.ObjectIdentifier{
+					GVK:        schema.GroupVersionKind{Group: cWObjRef.Group, Version: cWObjRef.Version, Kind: clusterWObj.GetObjectKind().GroupVersionKind().Kind},
+					Resource:   cWObjRef.Resource,
+					ObjectName: cache.NewObjectName("", cWObjRef.Name),
+				}
+				if v == util.BindingPolicyLabelSingletonStatusValueSet && len(bdg.Spec.Destinations) > 0 {
+					logger.V(2).Info("Singleton workload object should have status synced", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName)
+					c.singletonState.set(wObjIdentifier)
+				} else {
+					logger.V(2).Info("Singleton workload object status syncing is not driven by Binding", "resource", wObjIdentifier.Resource, "objectName", wObjIdentifier.ObjectName, "binding", bdg.Name)
+					c.singletonState.addIfNotExist(wObjIdentifier)
+				}
+			}
+		}
+	}
+
+	if !reconcile {
+		return nil
+	}
+
+	for wObjIdentifier := range c.singletonState.wObjSync {
```

Comment:
Data race here.

Reply:
Now the internal map was dropped, we don't have this problem. But still I fixed this before dropping.